src/join.rs

```diff
@@ -113,6 +129,97 @@ impl<J: Join> std::iter::Iterator for JoinIter<J> {
     }
 }
 
+/// `JoinParIter` is a `ParallelIterator` over a group of `Storages`.
+#[must_use]
+pub struct JoinParIter<J: Join>(J);
+
+impl<J> ParallelIterator for JoinParIter<J>
+  where
+    J: Join + Send,
+    J::Type: Send,
+    J::Value: Split + Send,
+    J::Mask: Send + Sync,
+{
+    type Item = J::Type;
+
+    fn drive_unindexed<C>(self, consumer: C) -> C::Result
+        where C: UnindexedConsumer<Self::Item>
+    {
+        let (keys, values) = self.0.open();
+        let producer = BitProducer((&keys).iter());
+        let values = UnsafeCell::new(values);
+        bridge_unindexed(JoinProducer::<J>::new(producer, &values), consumer)
+    }
+}
+
+struct JoinProducer<'a, J>
+  where
+    J: Join + Send,
+    J::Type: Send,
+    J::Value: 'a + Send,
+    J::Mask: 'a + Send + Sync,
+{
+    keys: BitProducer<'a, J::Mask>,
+    values: &'a UnsafeCell<J::Value>,
+}
+
+impl<'a, J> JoinProducer<'a, J>
+  where
+    J: Join + Send,
+    J::Type: Send,
+    J::Value: 'a + Send,
+    J::Mask: 'a + Send + Sync,
+{
+    fn new(keys: BitProducer<'a, J::Mask>, values: &'a UnsafeCell<J::Value>) -> Self {
+        JoinProducer {
+            keys: keys,
+            values: values,
+        }
+    }
+}
+
+unsafe impl<'a, J> Send for JoinProducer<'a, J>
+  where
+    J: Join + Send,
+    J::Type: Send,
+    J::Value: 'a + Send,
+    J::Mask: 'a + Send + Sync,
+{}
+
+impl<'a, J> UnindexedProducer for JoinProducer<'a, J>
+  where
+    J: Join + Send,
+    J::Type: Send,
+    J::Value: 'a + Send,
+    J::Mask: 'a + Send + Sync,
+{
+    type Item = J::Type;
+    fn split(self) -> (Self, Option<Self>) {
+        let (cur, other) = self.keys.split();
+        let prod = JoinProducer::new;
```

Comment:
This is a little confusing, could you just use `JoinProducer::new()` down there?